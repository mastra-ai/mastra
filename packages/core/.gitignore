*.log
.DS_Store
node_modules/
dist/
.env
*.d.ts
./memory
./vector/
./telemetry/
./storage/