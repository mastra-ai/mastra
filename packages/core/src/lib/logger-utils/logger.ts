import { Redis } from '@upstash/redis';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import path from 'path';

export const LogProvider = {
  CONSOLE: 'CONSOLE',
  FILE: 'FILE',
  UPSTASH: 'UPSTASH',
} as const;

export type LogConfig =
  | {
      provider: typeof LogProvider.CONSOLE;
      level?: LogLevel;
    }
  | {
      provider: typeof LogProvider.UPSTASH;
      level?: LogLevel;
      config: { url: string; token: string };
    }
  | {
      provider: typeof LogProvider.FILE;
      level?: LogLevel;
    };

export type LogProvider = (typeof LogProvider)[keyof typeof LogProvider];

export enum LogLevel {
  DEBUG,
  INFO,
  WARN,
  ERROR,
}

interface BaseLogMessage {
  message: string;
  destinationPath: string;
}

export interface Logger<T extends BaseLogMessage = BaseLogMessage> {
  debug(message: T | string, ...args: any[]): void;
  info(message: T | string, ...args: any[]): void;
  warn(message: T | string, ...args: any[]): void;
  error(message: T | string, ...args: any[]): void;
}

interface ConsoleLoggerOptions {
  level?: LogLevel;
}

interface FileLoggerOptions {
  dirPath: string;
  level?: LogLevel;
}

interface UpstashRedisLoggerOptions {
  redisClient: Redis;
  key: string;
  level?: LogLevel;
}

interface LoggerRegistry<T extends BaseLogMessage = BaseLogMessage> {
  [LogProvider.CONSOLE]: {
    logger: ConsoleLogger<T>;
    options: ConsoleLoggerOptions;
  };
  [LogProvider.FILE]: { logger: FileLogger<T>; options: FileLoggerOptions };
  [LogProvider.UPSTASH]: {
    logger: UpstashRedisLogger<T>;
    options: UpstashRedisLoggerOptions;
  };
}

// Abstract base class for loggers
abstract class BaseLogger<T extends BaseLogMessage = BaseLogMessage>
  implements Logger<T>
{
  protected level: LogLevel;

  constructor(level: LogLevel = LogLevel.INFO) {
    this.level = level;
  }

  abstract log(level: LogLevel, message: T | string, ...args: any[]): void;

  debug(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.DEBUG) {
      this.log(LogLevel.DEBUG, message, ...args);
    }
  }

  info(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.INFO) {
      this.log(LogLevel.INFO, message, ...args);
    }
  }

  warn(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.WARN) {
      this.log(LogLevel.WARN, message, ...args);
    }
  }

  error(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.ERROR) {
      this.log(LogLevel.ERROR, message, ...args);
    }
  }

  protected formatMessage(message: T | string): string {
    if (typeof message === 'string') {
      return message;
    }
    return JSON.stringify(message);
  }
}

/**
 * Logs to the console
 */
class ConsoleLogger<
  T extends BaseLogMessage = BaseLogMessage
> extends BaseLogger<T> {
  constructor({ level }: { level?: LogLevel }) {
    super(level ?? LogLevel.INFO);
  }

  log(level: LogLevel, message: T | string, ...args: any[]): void {
    const timestamp = new Date().toISOString();
    console.log(
      `[${timestamp}] [${level}] ${this.formatMessage(message)}`,
      ...args
    );
  }
}

/**
 * Logs to a file
 */
class FileLogger<
  T extends BaseLogMessage = BaseLogMessage
> extends BaseLogger<T> {
  #dirPath: string;

  constructor({ dirPath, level }: { dirPath: string; level?: LogLevel }) {
    super(level ?? LogLevel.INFO);
    if (!dirPath) {
      throw new Error('File path is required');
    }
    this.#dirPath = dirPath;
  }

  log(level: LogLevel, message: T): void {
    const fullPath = path.join(
      this.#dirPath,
      `${message.destinationPath}.json`
    );
    console.log(`Logging to file: ${fullPath}`);

    if (!existsSync(this.#dirPath)) {
      mkdirSync(this.#dirPath, { recursive: true });
      return writeFileSync(
        fullPath,
        JSON.stringify([{ ...message, level, createdAt: new Date() }], null, 2)
      );
    }

    if (!existsSync(fullPath)) {
      return writeFileSync(
        fullPath,
        JSON.stringify([{ ...message, createdAt: new Date() }], null, 2)
      );
    }

    const logs = JSON.parse(readFileSync(fullPath, 'utf-8'));
    logs.push({ ...message, createdAt: new Date() });

    return writeFileSync(fullPath, JSON.stringify(logs, null, 2));
  }
}

/**
 * Logs to Upstash Redis
 */
class UpstashRedisLogger<
  T extends BaseLogMessage = BaseLogMessage
> extends BaseLogger<T> {
  #redisClient: Redis;
  #key: string;

  constructor({
    redisClient,
    level,
    key,
  }: {
    redisClient: Redis;
    level?: LogLevel;
    key: string;
  }) {
    super(level ?? LogLevel.INFO);
    if (!redisClient) {
      throw new Error('redisClient is required');
    }
    if (!key) {
      throw new Error('Redis storage key is required');
    }
    this.#redisClient = redisClient;
    this.#key = key;
  }

  log(level: LogLevel, message: T): void {
    const fullKey = path.join(this.#key, `${message.destinationPath}`);
    this.#redisClient.lpush(
      fullKey,
      JSON.stringify({ ...message, level, createdAt: new Date() })
    );
  }

  getLogs(key: string) {
    return this.#redisClient.lrange(key, 0, -1);
  }
}

export function createLogger<
  T extends BaseLogMessage = BaseLogMessage,
  K extends keyof LoggerRegistry<T> = keyof LoggerRegistry<T>
>({
  type,
  options,
}: {
  type: K;
  options?: LoggerRegistry[K]['options'];
}): LoggerRegistry<T>[K]['logger'] {
  switch (type) {
    case 'CONSOLE':
      return new ConsoleLogger<T>({ level: options?.level });
    case 'FILE':
      return new FileLogger<T>({
        dirPath: (options as FileLoggerOptions).dirPath,
        level: (options as FileLoggerOptions).level,
      });
    case 'UPSTASH':
      return new UpstashRedisLogger<T>({
        redisClient: (options as UpstashRedisLoggerOptions).redisClient,
        key: (options as UpstashRedisLoggerOptions).key,
        level: (options as UpstashRedisLoggerOptions).level,
      });
    default:
      throw new Error(`Unsupported logger type: ${type}`);
  }
}
