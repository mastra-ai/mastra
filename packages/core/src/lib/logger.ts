import { Redis } from "@upstash/redis";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import path from "path";

const LogLevel = {
	DEBUG: 'DEBUG',
	INFO: 'INFO',
	WARN: 'WARN',
	ERROR: 'ERROR',
} as const;

export type LogLevel = typeof LogLevel[keyof typeof LogLevel];

interface BaseLogMessage {
  message: string;
  destinationPath: string;
}

interface Logger<T extends BaseLogMessage = BaseLogMessage> {
  debug(message: T | string, ...args: any[]): void;
  info(message: T | string, ...args: any[]): void;
  warn(message: T | string, ...args: any[]): void;
  error(message: T | string, ...args: any[]): void;
}

// Abstract base class for loggers
abstract class BaseLogger<T extends BaseLogMessage = BaseLogMessage> implements Logger<T> {
  protected level: LogLevel;

  constructor(level: LogLevel = LogLevel.INFO) {
    this.level = level;
  }

  abstract log(level: LogLevel, message: T | string, ...args: any[]): void;

  debug(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.DEBUG) {
      this.log(LogLevel.DEBUG, message, ...args);
    }
  }

  info(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.INFO) {
      this.log(LogLevel.INFO, message, ...args);
    }
  }

  warn(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.WARN) {
      this.log(LogLevel.WARN, message, ...args);
    }
  }

  error(message: T | string, ...args: any[]): void {
    if (this.level <= LogLevel.ERROR) {
      this.log(LogLevel.ERROR, message, ...args);
    }
  }

  protected formatMessage(message: T | string): string {
    if (typeof message === 'string') {
      return message;
    }
    return JSON.stringify(message);
  }
}

/**
 * Logs to the console
 */
class ConsoleLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  log(level: LogLevel, message: T | string, ...args: any[]): void {
    const timestamp = new Date().toISOString();
    console.log(`[${timestamp}] [${level}] ${this.formatMessage(message)}`, ...args);
  }
}

/**
 * Logs to a file
 */
class FileLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  #filePath: string;

  constructor(filePath: string, level: LogLevel = LogLevel.INFO) {
    super(level);
    this.#filePath = filePath;
  }

  log(level: LogLevel, message: T): void {
		const fullPath = path.join(this.#filePath, message.destinationPath);
    // Implement file logging logic here
    console.log(`Logging to file: ${fullPath}`);

		if (!existsSync(this.#filePath)) {
			mkdirSync(this.#filePath, { recursive: true });
			return writeFileSync(
				fullPath,
				JSON.stringify([{ ...message, level, createdAt: new Date() }], null, 2)
			);
		}

		if (!existsSync(fullPath)) {
			return writeFileSync(
				fullPath,
				JSON.stringify([{ ...message, createdAt: new Date() }], null, 2)
			);
		}

		const logs = JSON.parse(readFileSync(fullPath, 'utf-8'));
		logs.push({ ...message, createdAt: new Date() });

		return writeFileSync(fullPath, JSON.stringify(logs, null, 2));
  }
}

/**
 * Logs to Upstash Redis
 */
class UpstashRedisLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  #redisClient: Redis;
  #key: string;

  constructor(redisClient: Redis, level: LogLevel = LogLevel.INFO, key: string) {
    super(level);
    this.#redisClient = redisClient;
    this.#key = key;
  }

  log(level: LogLevel, message: T): void {
    this.#redisClient.lpush(this.#key, JSON.stringify({...message, level, createdAt: new Date()}));
  }

	getLogs() {
		return this.#redisClient.lrange(this.#key, 0, -1);
	}
}

export function createLogger<T extends BaseLogMessage = BaseLogMessage>(
  type: 'CONSOLE' | 'FILE' | 'UPSTASH',
  options?: any
): Logger<T> {
  switch (type) {
    case 'CONSOLE':
      return new ConsoleLogger<T>(options?.level);
    case 'FILE':
      return new FileLogger<T>(options?.filePath, options?.level);
    case 'UPSTASH':
      return new UpstashRedisLogger<T>(options?.redisClient, options?.level, options?.key);
    default:
      throw new Error(`Unsupported logger type: ${type}.\n\tExpected CONSOLE, FILE or UPSTASH`);
  }
}
