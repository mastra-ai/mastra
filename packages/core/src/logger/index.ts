import { Redis } from '@upstash/redis';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import path from 'path';

import { Run } from '../run/types';

// Constants and Types
export const RegisteredLogger = {
  AGENT: 'AGENT',
  WORKFLOW: 'WORKFLOW',
  LLM: 'LLM',
  TTS: 'TTS',
} as const;

export type RegisteredLogger = (typeof RegisteredLogger)[keyof typeof RegisteredLogger];

export const LogLevel = {
  DEBUG: 'DEBUG',
  INFO: 'INFO',
  WARN: 'WARN',
  ERROR: 'ERROR',
  NONE: 'NONE',
} as const;

export type LogLevel = (typeof LogLevel)[keyof typeof LogLevel];

// Base Interfaces
export interface BaseLogMessage extends Run {
  message: string;
  destinationPath: string;
  type: RegisteredLogger;
}

export interface Logger<T extends BaseLogMessage = BaseLogMessage> {
  debug(message: T | string, ...args: any[]): void | Promise<void>;
  info(message: T | string, ...args: any[]): void | Promise<void>;
  warn(message: T | string, ...args: any[]): void | Promise<void>;
  error(message: T | string, ...args: any[]): void | Promise<void>;
  cleanup?(): Promise<void>;
}

type ConsoleLoggerConfig = { type: 'CONSOLE'; level?: LogLevel };
type FileLoggerConfig = { type: 'FILE'; level?: LogLevel; dirPath?: string };
type UpstashLoggerConfig = {
  type: 'UPSTASH';
  level?: LogLevel;
  url: string;
  token: string;
  key?: string;
};

type LoggerConfig = ConsoleLoggerConfig | FileLoggerConfig | UpstashLoggerConfig;

type LoggerTypeMap = {
  CONSOLE: ConsoleLogger<BaseLogMessage>;
  FILE: FileLogger<BaseLogMessage>;
  UPSTASH: UpstashRedisLogger<BaseLogMessage>;
};

// Abstract Base Logger
export abstract class BaseLogger<T extends BaseLogMessage = BaseLogMessage> implements Logger<T> {
  protected level: LogLevel;

  constructor(level: LogLevel = LogLevel.INFO) {
    this.level = level;
  }

  abstract log(level: LogLevel, message: T | string, ...args: any[]): void | Promise<void>;

  debug(message: T | string, ...args: any[]): void | Promise<void> {
    if (this.level === LogLevel.DEBUG) {
      return this.log(LogLevel.DEBUG, message, ...args);
    }
  }

  info(message: T | string, ...args: any[]): void | Promise<void> {
    if ([LogLevel.INFO, LogLevel.DEBUG, LogLevel.ERROR].includes(this.level as any) && this.level !== LogLevel.NONE) {
      return this.log(LogLevel.INFO, message, ...args);
    }
  }

  warn(message: T | string, ...args: any[]): void | Promise<void> {
    if ([LogLevel.WARN, LogLevel.ERROR, LogLevel.DEBUG].includes(this.level as any) && this.level !== LogLevel.NONE) {
      return this.log(LogLevel.WARN, message, ...args);
    }
  }

  error(message: T | string, ...args: any[]): void | Promise<void> {
    if (
      [LogLevel.ERROR, LogLevel.INFO, LogLevel.WARN, LogLevel.DEBUG].includes(this.level as any) &&
      this.level !== LogLevel.NONE
    ) {
      return this.log(LogLevel.ERROR, message, ...args);
    }
  }

  protected formatMessage(message: T | string): string {
    if (typeof message === 'string') {
      return message;
    }
    return JSON.stringify(message);
  }

  protected formatLogEntry(level: LogLevel, message: T | string): any {
    return {
      timestamp: new Date().toISOString(),
      level: LogLevel[level],
      message: this.formatMessage(message),
    };
  }

  async getLogsByRunId(runId: string): Promise<T[]> {
    console.warn(`getLogsByRunId ${runId} not implemented for ${this.constructor.name}`);
    return [];
  }

  async getLogs(): Promise<string[]> {
    console.warn(`getLogs not implemented for ${this.constructor.name}`);
    return [];
  }
}

// Console Logger Implementation
export class ConsoleLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  private logs: { timestamp: string; level: string; message: string; runId?: string }[] = [];
  // private originalConsoleLog: typeof console.log;

  constructor(level?: LogLevel) {
    super(level ?? LogLevel.INFO);
    // Store original console.log and bind it
    // this.originalConsoleLog = console.log.bind(console);
  }

  log(level: LogLevel, message: T | string, ...args: any[]): void {
    let runId: string | undefined;
    if (typeof message !== 'string') {
      runId = message.runId;
    }

    const logEntry = this.formatLogEntry(level, message);
    const logMessage = `[${logEntry.timestamp}] [${logEntry.level}] ${logEntry.message}`;

    // Store the log message
    this.logs.push({ ...logEntry, runId });

    console.log(logMessage, ...args);
  }

  async getLogs(): Promise<string[]> {
    return this.logs?.map(log => `[${log.timestamp}] [${log.level}] ${log.message}`) || [];
  }

  async getLogsByRunId(runId: string): Promise<T[]> {
    return (this.logs?.filter(log => log.runId === runId) as unknown as T[]) || [];
  }
}

// File Logger Implementation
export class FileLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  #dirPath: string;

  constructor(dirPath: string = 'logs', level?: LogLevel) {
    super(level ?? LogLevel.INFO);
    this.#dirPath = dirPath;
  }

  log(level: LogLevel, message: T): void {
    if (typeof message === 'string') {
      throw new Error('FileLogger requires a BaseLogMessage object');
    }

    const fullPath = path.join(this.#dirPath, `${message.destinationPath}.json`);

    this.ensureDirectoryExists();
    this.writeLogToFile(fullPath, level, message);
  }

  private ensureDirectoryExists(): void {
    if (!existsSync(this.#dirPath)) {
      mkdirSync(this.#dirPath, { recursive: true });
    }
  }

  private writeLogToFile(fullPath: string, level: LogLevel, message: T): void {
    const logEntry = {
      ...message,
      level: LogLevel[level],
      createdAt: new Date(),
    };

    let logs = [];
    if (existsSync(fullPath)) {
      logs = JSON.parse(readFileSync(fullPath, 'utf-8'));
    }

    logs.push(logEntry);
    writeFileSync(fullPath, JSON.stringify(logs, null, 2));
  }
}

// Upstash Redis Logger Implementation
export class UpstashRedisLogger<T extends BaseLogMessage = BaseLogMessage> extends BaseLogger<T> {
  #redis: Redis;
  #key: string;

  constructor(redis: Redis, key: string = 'logs', level?: LogLevel) {
    super(level ?? LogLevel.INFO);
    this.#redis = redis;
    this.#key = key;
  }

  async log(level: LogLevel, message: T): Promise<void> {
    if (typeof message === 'string') {
      throw new Error('UpstashRedisLogger requires a BaseLogMessage object');
    }

    const logEntry = {
      ...message,
      level: LogLevel[level],
      createdAt: new Date(),
      runId: message.runId,
    };

    const runKey = `${this.#key}:run:${message.runId}`;

    if (message.runId) {
      await Promise.all([
        this.#redis.lpush(this.#key, JSON.stringify(logEntry)),
        this.#redis.lpush(runKey, JSON.stringify(logEntry)),
      ]);
    } else {
      await this.#redis.lpush(this.#key, JSON.stringify(logEntry));
    }
  }

  async getLogs(): Promise<string[]> {
    return this.#redis.lrange(this.#key, 0, -1);
  }

  async deleteLogsByKey(key: string): Promise<void> {
    await this.#redis.del(key);
  }

  async getLogsByRunId(runId: string): Promise<T[]> {
    if (!runId) {
      throw new Error('runId is required');
    }

    try {
      const runKey = `${this.#key}:run:${runId}`;
      const logs = await this.#redis.lrange(runKey, 0, -1);

      return logs.reduce((acc: T[], logStr: string) => {
        try {
          const log = typeof logStr === 'string' ? JSON.parse(logStr) : logStr;
          if (log && typeof log === 'object' && log.runId === runId) {
            acc.push(log as T);
          }
        } catch (parseError) {
          if (typeof logStr === 'string') {
            console.error(`Failed to parse log entry: ${logStr}`, parseError);
          }
        }
        return acc;
      }, []);
    } catch (error) {
      console.error(`Failed to fetch logs for runId ${runId}:`, error);
      return [];
    }
  }

  async cleanup(): Promise<void> {
    // Cleanup Redis connection if needed
  }
}

// Multi Logger Implementation
export class MultiLogger<T extends BaseLogMessage = BaseLogMessage> implements Logger<T> {
  private loggers: Logger<T>[];

  constructor(loggers: Logger<T>[]) {
    this.loggers = loggers;
  }

  async debug(message: T | string, ...args: any[]): Promise<void> {
    await Promise.all(this.loggers.map(logger => logger.debug(message, ...args)));
  }

  async info(message: T | string, ...args: any[]): Promise<void> {
    await Promise.all(this.loggers.map(logger => logger.info(message, ...args)));
  }

  async warn(message: T | string, ...args: any[]): Promise<void> {
    await Promise.all(this.loggers.map(logger => logger.warn(message, ...args)));
  }

  async error(message: T | string, ...args: any[]): Promise<void> {
    await Promise.all(this.loggers.map(logger => logger.error(message, ...args)));
  }

  async cleanup(): Promise<void> {
    await Promise.all(
      this.loggers.map(async logger => {
        if (logger.cleanup) {
          await logger.cleanup();
        }
      }),
    );
  }
}

// Factory function for built-in loggers
// In createLogger function
export const createLogger = <Type extends LoggerConfig['type'], T extends BaseLogMessage = BaseLogMessage>(
  config: Extract<LoggerConfig, { type: Type }>,
): LoggerTypeMap[Type] => {
  switch (config.type) {
    case 'CONSOLE':
      return new ConsoleLogger<T>(config.level) as unknown as LoggerTypeMap[Type];
    case 'FILE': {
      const fileConfig = config as FileLoggerConfig;
      return new FileLogger<T>(fileConfig.dirPath, fileConfig.level) as unknown as LoggerTypeMap[Type];
    }
    case 'UPSTASH': {
      const upstashConfig = config as UpstashLoggerConfig;
      const redis = new Redis({
        url: upstashConfig.url,
        token: upstashConfig.token,
      });
      return new UpstashRedisLogger<T>(redis, upstashConfig.key, upstashConfig.level) as unknown as LoggerTypeMap[Type];
    }
    default: {
      const exhaustiveCheck: never = config.type;
      throw new Error(`Unsupported logger type: ${exhaustiveCheck}`);
    }
  }
};

export function combineLoggers<T extends BaseLogMessage = BaseLogMessage>(loggers: Logger<T>[]): Logger<T> {
  return new MultiLogger<T>(loggers);
}

export const noopLogger = {
  debug: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
  cleanup: () => {},
};
