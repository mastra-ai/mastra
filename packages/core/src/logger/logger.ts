import { LogLevel } from './constants';
import type { LoggerTransport } from './transport';

export interface IMastraLogger {
  debug(message: string, ...args: any[]): void;
  info(message: string, ...args: any[]): void;
  warn(message: string, ...args: any[]): void;
  error(message: string, ...args: any[]): void;

  getTransports(): Map<string, LoggerTransport>;
  getLogs(_transportId: string): Promise<any[]>;
  getLogsByRunId(_args: { transportId: string; runId: string }): Promise<any[]>;
}

export abstract class MastraLogger implements IMastraLogger {
  protected name: string;
  protected level: LogLevel;
  protected transports: Map<string, LoggerTransport>;

  constructor(
    options: {
      name?: string;
      level?: LogLevel;
      transports?: Record<string, LoggerTransport>;
    } = {},
  ) {
    this.name = options.name || 'Mastra';
    this.level = options.level || LogLevel.ERROR;
    this.transports = new Map(Object.entries(options.transports || {}));
  }

  abstract debug(message: string, ...args: any[]): void;
  abstract info(message: string, ...args: any[]): void;
  abstract warn(message: string, ...args: any[]): void;
  abstract error(message: string, ...args: any[]): void;

  getTransports() {
    return this.transports;
  }

  async getLogs(transportId: string) {
    if (!transportId || !this.transports.has(transportId)) {
      return [];
    }

    return this.transports.get(transportId)!.getLogs() ?? [];
  }

  async getLogsByRunId({ transportId, runId }: { transportId: string; runId: string }) {
    if (!transportId || !this.transports.has(transportId) || !runId) {
      return [];
    }

    return this.transports.get(transportId)!.getLogsByRunId({ runId }) ?? [];
  }
}
