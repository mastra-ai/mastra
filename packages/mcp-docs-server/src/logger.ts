import type { Server } from '@modelcontextprotocol/sdk/server/index.js';
import type { LoggingLevel } from '@modelcontextprotocol/sdk/types.js';

// Logger interface for type safety
export interface Logger {
  info: (message: string, data?: any) => Promise<void>;
  warning: (message: string, data?: any) => Promise<void>;
  error: (message: string, error?: any) => Promise<void>;
  debug: (message: string, data?: any) => Promise<void>;
}

export function convertLogLevelToLoggerMethod(level: LoggingLevel): 'debug' | 'info' | 'warn' | 'error' {
  switch (level) {
    case 'debug':
      return 'debug';
    case 'info':
    case 'notice':
      return 'info';
    case 'warning':
      return 'warn';
    case 'error':
    case 'critical':
    case 'alert':
    case 'emergency':
      return 'error';
    default:
      // For any other levels, default to info
      return 'info';
  }
}

// Create logger factory to inject server instance
export function createLogger(server?: Server): Logger {
  // const writeLog = (level: string, message: string, data?: any) => {
  //   const logMessage = {
  //     jsonrpc: '2.0',
  //     method: 'notifications/tools',
  //     params: {
  //       level,
  //       message,
  //       ...(data ? (typeof data === 'object' ? data : { data }) : {}),
  //     },
  //   };
  //   process.stdout.write(JSON.stringify(logMessage) + '\n');
  // };

  const sendLog = async (level: 'error' | 'debug' | 'info' | 'warning', message: string, data?: any) => {
    if (!server) return;

    try {
      await server.sendLoggingMessage({
        level,
        data: {
          message,
          ...(data ? (typeof data === 'object' ? data : { data }) : {}),
        },
      });
    } catch (error) {
      if (
        error instanceof Error &&
        (error.message === 'Not connected' ||
          error.message.includes('does not support logging') ||
          error.message.includes('Connection closed'))
      ) {
        return;
      }
      console.error(`Failed to send ${level} log:`, error instanceof Error ? error.message : error);
    }
  };

  return {
    info: async (message: string, data?: any) => {
      // writeLog('info', message, data);
      await sendLog('info', message, data);
    },
    warning: async (message: string, data?: any) => {
      // writeLog('warning', message, data);
      await sendLog('warning', message, data);
    },
    error: async (message: string, error?: any) => {
      const errorData =
        error instanceof Error
          ? {
              message: error.message,
              stack: error.stack,
              name: error.name,
            }
          : error;
      // writeLog('error', message, errorData);
      await sendLog('error', message, errorData);
    },
    debug: async (message: string, data?: any) => {
      if (process.env.DEBUG || process.env.NODE_ENV === 'development') {
        // writeLog('debug', message, data);
        await sendLog('debug', message, data);
      }
    },
  };
}

// Create a default logger instance
export const logger = createLogger();
