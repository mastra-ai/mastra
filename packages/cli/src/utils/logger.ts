import { createLogger } from '@mastra/core/logger';

export const logger = createLogger({
  name: 'Mastra CLI',
  level: 'debug',
});
