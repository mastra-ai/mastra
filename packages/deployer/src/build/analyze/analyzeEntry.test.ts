import { analyzeEntry } from './analyzeEntry';
import { describe, it, expect, vi } from 'vitest';
import { readFile } from 'fs-extra';
import { join } from 'path';
import { noopLogger } from '@mastra/core/logger';

vi.spyOn(process, 'cwd').mockReturnValue(join(import.meta.dirname, '__fixtures__', 'default'));

describe('analyzeEntry', () => {
  it('should analyze the entry file', async () => {
    const entryAsString = await readFile(join(import.meta.dirname, '__fixtures__', 'default', 'entry.ts'), 'utf-8');

    const result = await analyzeEntry({ entry: entryAsString, isVirtualFile: true }, ``, {
      logger: noopLogger,
      sourcemapEnabled: false,
      workspaceMap: new Map(),
    });

    expect(result.dependencies.size).toBe(4);

    // Check individual dependencies without hardcoded paths
    expect(result.dependencies.has('@mastra/core/logger')).toBe(true);
    expect(result.dependencies.has('@mastra/core/mastra')).toBe(true);
    expect(result.dependencies.has('@mastra/core/agent')).toBe(true);
    expect(result.dependencies.has('@ai-sdk/openai')).toBe(true);

    const loggerDep = result.dependencies.get('@mastra/core/logger');
    expect(loggerDep?.exports).toEqual(['createLogger']);
    expect(loggerDep?.isWorkspace).toBe(false);
    expect(loggerDep?.rootPath).toMatch(/packages\/core$/);

    const mastraDep = result.dependencies.get('@mastra/core/mastra');
    expect(mastraDep?.exports).toEqual(['Mastra']);
    expect(mastraDep?.isWorkspace).toBe(false);
    expect(mastraDep?.rootPath).toMatch(/packages\/core$/);

    const agentDep = result.dependencies.get('@mastra/core/agent');
    expect(agentDep?.exports).toEqual(['Agent']);
    expect(agentDep?.isWorkspace).toBe(false);
    expect(agentDep?.rootPath).toMatch(/packages\/core$/);

    const openaiDep = result.dependencies.get('@ai-sdk/openai');
    expect(openaiDep?.exports).toEqual(['openai']);
    expect(openaiDep?.isWorkspace).toBe(false);
    expect(openaiDep?.rootPath).toBe(null);

    expect(result.output).toMatchSnapshot();
  });
});
