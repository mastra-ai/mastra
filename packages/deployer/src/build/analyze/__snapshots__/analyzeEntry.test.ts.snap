// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyzeEntry > should analyze the entry file 1`] = `
{
  "code": "import { createLogger } from '@mastra/core/logger';
import { Mastra } from '@mastra/core/mastra';
import { Agent } from '@mastra/core/agent';
import { openai } from '@ai-sdk/openai';

const testAgent = new Agent({
  name: "test-agent",
  instructions: "test-agent",
  model: openai("gpt-4o")
});

const mastra = new Mastra({
  agents: { testAgent },
  logger: createLogger({
    name: 'Mastra',
    level: 'info',
  }),
});

export { mastra };
",
  "map": null,
}
`;
